unreal/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+//! API to process Unreal Engine 4 crashes
+#![warn(missing_docs)]
+
+extern crate byteorder;
+extern crate bytes;
+extern crate compress;
+extern crate failure;
+
+use std::io::{self, Cursor, Read};
+
+use bytes::{Buf, Bytes};
+use compress::zlib;
+use failure::Fail;
+
+struct Header {
+    pub directory_name: String,
+    pub file_name: String,
+    pub uncompressed_size: i32,
+    pub file_count: i32,
+}
+
+/// Meta-data about a file within a UE4 crash file
+#[derive(Debug)]
+pub struct CrashFileMeta {
+    /// The original index within the UE4 crash file
+    pub index: usize,
+    /// File name
+    pub file_name: String,
+    /// Start of the file within crash dumb
+    pub offset: usize,
+    /// Length of bytes from offset
+    pub len: usize,
+}
+
+/// Errors related to parsing an UE4 crash file
+#[derive(Fail, Debug)]
+pub enum Unreal4ParseError {
+    /// Expected UnrealEngine4 crash (zlib compressed)
+    #[fail(display = "unknown bytes format")]
+    UnknownBytesFormat,
+    /// Value out of bounds
+    #[fail(display = "out of bounds")]
+    OutOfBounds,
+    /// Invalid compressed data
+    #[fail(display = "bad compression")]
+    BadCompression(io::Error),
+}
+
+/// Unreal Engine 4 crash file
+#[derive(Debug)]
+pub struct Unreal4Crash {
+    bytes: Bytes,
+    files: Vec<CrashFileMeta>,
+}
+
+impl Unreal4Crash {
+    /// Creates an instance of `Unreal4Crash` from the original, compressed bytes
+    pub fn from_bytes(bytes: &[u8]) -> Result<Unreal4Crash, Unreal4ParseError> {
```

Comment:
This should be called `from_slice`. Also, it would be cool to have two functions:
 - `from_slice`: Assumes already decompressed
 - `from_slice_compressed`: Decompresses and calls `from_slice`.

That way we have more flexibility when calling this from Sentry.

Reply:
Agree to rename `from_slice`.

I'm not sure it's worth adding the extra API "just in case" since I can't see us doing the zlib decompressing in python to pass it to symbolic and do nothing else with the blob. The logic to parse the blob is in symbolic so doesn't make much sense I think.

If for some reason we end up doing that, we could add the API here later.

Reply:
Then please rename to `_compressed` right away.

Reply:
an Unreal Engine 4 crash is compressed.
IMO there's no need to say: from_slice_zlib_compressed_etc since it's simply give it the cras as-is.

Reply:
Alright, `from_slice` wfm 👍